File: in_Java/0106 Construct Binary Tree from Inorder and Postorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        HashMap<Integer,Integer> hm = new HashMap<>();
        for(int i = 0; i < inorder.length; i++) hm.put(inorder[i],i);
        TreeNode r = helper(inorder, 0, inorder.length-1, postorder, 0, postorder.length-1, hm);
        return r;
    }
    private TreeNode helper(int[] in, int inStart, int inEnd, int[] post, int postStart, int postEnd, HashMap<Integer,Integer> hm){
        if(inStart > inEnd || postStart > postEnd) return null;
        TreeNode root = new TreeNode(post[postEnd]);
        int id = hm.get(root.val);
        int dis = inEnd - id;
        root.right = helper(in, id+1, inEnd, post, postEnd-dis, postEnd-1, hm);
        root.left = helper(in, inStart, id-1, post, postStart, postEnd-dis-1, hm);
        return root;
    }
}

File: in_Python/0106 Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        d = {}
        for i,x in enumerate(inorder):
            d[x] = i
        def recur(l,r):
            if l > r:
                return None
            m = TreeNode(postorder.pop())
            idx = d[m.val]
            m.right = recur(idx+1, r)
            m.left = recur(l,idx-1)
            return m
        return recur(0, len(inorder)-1)
            
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        if inorder:
            r = postorder.pop()
            idx = -1
            for i,x in enumerate(inorder):
                if x == r:
                    idx = i
                    break
            ro = TreeNode(r)
            
            ro.right = self.buildTree(inorder[idx+1:], postorder)
            ro.left = self.buildTree(inorder[:idx], postorder)
            return ro